File: search/binary_search.py
# 재귀 함수를 이용한 이진 탐색

def binary_search(array,target,start,end) : 
    if start>end :
        return None
    mid = (start+end)//2

    if array[mid] == target : 
        return mid+1    # 결과는 n 번쨰에 있다.
    if array[mid] > target : 
        return binary_search(array,target,start,mid-1)
    elif array[mid] < target : 
        return binary_search(array,target,mid+1,end)

array = [0,2,4,6,8,10,12,14,16,18]

print(binary_search(array,4,0,len(array)-1))

File: search/BinarySearch.java
public class BinarySearch {

    // 이진 탐색
    public static int binarySearch() {

        int[] array = new int[] {1,4,7,11,15,21,24,27,30};
        int start = 0;
        int end = array.length - 1;
        int target = 11;
        while(start <= end) {

            int mid = (start + end) / 2;

            if (array[mid] == target) {
                return mid;
            } else if (array[mid] > target) {
                end = mid - 1;
            } else {
                start = mid + 1;
            }
        }

        return -1;
    }
}
